### PR TITLE
III-7190 matching birthdate range resultset fixes

### PR DESCRIPTION
**Pull request by JonasVHG**
Added

  - New top-level matchingBirthdateRanges field on offer search responses: when the query contains a birthdate range (via advanced q birthdateRange:[from TO to]
  clauses — flat, grouped or repeated — or the structured birthdateRangeFrom/birthdateRangeTo params), the response includes one entry per queried range with its
  from, to and the list of returned event @ids that qualify for it. Omitted when no birthdate range is queried.
  - BirthdateRangeQueryStringParser: single home for the birthdateRange: query-string patterns; parses a q string into BirthdateRange[].
  - MatchingBirthdateRangesResolver: resolves the queried ranges from a request and computes per-range matches, mirroring createBirthdateRangeQuery (indexed
  birthdateRange intersects the queried dates, or typicalAgeRange intersects the equivalent age range excluding all-ages events).
  - ElasticSearchOfferQueryBuilder::withBirthdateRangeMatchFields(): widens _source with birthdateRange/typicalAgeRange/allAges, applied only when a birthdate
  range is queried.

  Changed

  - OfferSearchController resolves queried birthdate ranges after request parsing (widening _source when needed) and attaches matchingBirthdateRanges to the
  response next to the existing facet output; wired via OfferSearchControllerFactory.
  - BirthdateRangeToTypicalAgeRangeQueryStringFactory now reuses the shared BirthdateRangeQueryStringParser patterns instead of its own private regex constants
  (behaviour unchanged).

  Removed
  
  - Duplicate birthdateRange: regex constants from BirthdateRangeToTypicalAgeRangeQueryStringFactory (moved into BirthdateRangeQueryStringParser).

  Related PR

  - https://github.com/cultuurnet/udb3-backend/pull/2325

  ---
  Ticket: https://jira.publiq.be/browse/III-7190

### CODE REVIEW

**Comment by JonasVHG** [comment]
@claude review.

File: src/Http/Offer/MatchingBirthdateRangesResolver.php [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        } catch (UnsupportedParameterValue $e) {
+            return null;
+        }
+    }
+
+    /**
+     * @param BirthdateRange[] $ranges
+     * @return BirthdateRange[]
+     */
+    private function deduplicate(array $ranges): array
```
Comment (grubolsch) [inline src/Http/Offer/MatchingBirthdateRangesResolver.php:106]: Just food for thought, we should maybe look if we can port the collection classes we have in UDB3 Backend (CultuurNet\UDB3\Model\ValueObject\Collection) to SAPI3, so we can jsut pass collection objects, that would make this code more readable.
You don't need to change this now

File: tests/Http/OfferSearchControllerTest.php [hunk 2]
```diff
@@ -380,6 +392,76 @@ public function it_returns_a_paged_collection_of_search_results_based_on_request
         $this->assertEquals($expectedJsonResponse, $actualJsonResponse);
     }
 
+    /**
+     * @test
+     */
+    public function it_adds_matching_birthdate_ranges_when_a_birthdate_range_is_queried(): void
```
Comment (grubolsch) [inline tests/Http/OfferSearchControllerTest.php:398]: Maybe we should add a check that matchingBirthdateRanges is missing when you do not query a birthdate range. Probably already implicit tested somewhere else but good to make a formal contract about it.

File: tests/ElasticSearch/BirthdateRangeToTypicalAgeRangeQueryStringFactoryTest.php [hunk 3]
```diff
@@ -67,13 +67,93 @@ public function it_leaves_a_query_without_a_birthdate_range_unchanged(): void
...
+            . ' AND name.nl:foo'
+        );
+
+        $this->assertEquals($expected, $actual);
+    }
+
+    /**
+     * @test
+     */
+    public function it_skips_an_invalid_range_inside_a_grouped_birthdate_range(): void
```
Comment (grubolsch) [inline tests/ElasticSearch/BirthdateRangeToTypicalAgeRangeQueryStringFactoryTest.php:135]: Good test, but should we also have a test where every subrange is invalid? There is a specific if in the code for this edge case that could be tested:

                if ($ageClauses === []) {
                    return $original;
                }

File: src/Http/Offer/MatchingBirthdateRangesResolver.php [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+ * which of the returned events qualify for it, so a consumer can show per-child "geschikt voor"
+ * hints when multiple children are searched in one query.
+ *
+ * The match test mirrors ElasticSearchOfferQueryBuilder::createBirthdateRangeQuery(): an event
+ * qualifies when its indexed `birthdateRange` intersects the queried date range, or when its
+ * `typicalAgeRange` intersects the equivalent age range (excluding "all ages" events). The
+ * birthdate <-> age conversion is relative to "now", matching the query-time behaviour everywhere
+ * else in the codebase.
+ */
+final class MatchingBirthdateRangesResolver
```
Comment (grubolsch) [inline src/Http/Offer/MatchingBirthdateRangesResolver.php:24]: I think this comment is very verbose, a lot of this can just be read from the code, focus more on the why?
Here is a suggestion how to shorten it:

 * Resolves which queried birthdate ranges match each returned event.
 *
 * Supports both `q=birthdateRangeFrom:[from TO to]` and
 * `birthdateRangeFrom`/`birthdateRangeTo` parameters. An event matches if its
 * `birthdateRange` or (non-"all ages") `typicalAgeRange` intersects the
 * queried range, using the same "now"-relative logic as the search query.
 */

**Comment by grubolsch** [comment]
If I understand correctly `documentMatchesRange()` and
`createBirthdateRangeQuery()` kinda do the same thing, but they are on
different levels (post-fetch PHP check on already-returned documents vs.
the actual ES filter query that decides the result set in the first
place). 
My only problem is that they have to keep matching to the same
result set, or in the future we risk changing the boundaries in one of
the two, but not the other, right now the only thing keeping them in
sync is a doc-comment ("mirrors ..."), nothing fails if one changes
without the other. Since this drives a public API field, a silent drift
here would be a hard-to-notice correctness bug (e.g. `matchingBirthdateRanges`
claiming a match for an event that isn't even in the result set, or vice
versa).

Can we maybe make a test that starts with the same fixture and checks if they return the same boundries?

**Review by grubolsch — CHANGES_REQUESTED** [review]

File: src/Http/Offer/MatchingBirthdateRangesResolver.php [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        $to = isset($birthdateRange->lte) ? DateTimeImmutable::createFromFormat('!Y-m-d', (string) $birthdateRange->lte) : null;
+        $from = $from instanceof DateTimeImmutable ? $from : null;
+        $to = $to instanceof DateTimeImmutable ? $to : null;
+
+        if ($from === null && $to === null) {
+            return false;
+        }
+
+        // Intersection with [range->from, range->to]; a missing bound is unbounded.
+        if ($from !== null && $from > $range->getTo()) {
```
Comment (grubolsch) [inline src/Http/Offer/MatchingBirthdateRangesResolver.php:144]: Comment does not add much

File: tests/ElasticSearch/BirthdateRangeToTypicalAgeRangeQueryStringFactoryTest.php (hunk 3 above)
Comment (JonasVHG) [reply tests/ElasticSearch/BirthdateRangeToTypicalAgeRangeQueryStringFactoryTest.php:135]: Done in previous PR.

File: src/Http/Offer/MatchingBirthdateRangesResolver.php (hunk 1 above)
Comment (JonasVHG) [reply src/Http/Offer/MatchingBirthdateRangesResolver.php:106]: Indeed, worth creating a ticket for it or mentioning it in a weekly.

Comment (grubolsch) [reply src/Http/Offer/MatchingBirthdateRangesResolver.php:106]: Mentioned it in the biweekly notes.

**Comment by grubolsch** [comment]
I run the jenkins tests and then I will merge.

https://jenkins.publiq.be/job/uitdatabank-acceptance-tests/823/

**Comment by grubolsch** [comment]
@claude

**Review by grubolsch — APPROVED** [review]
